src/main/haskell/kore/src/Kore/Builtin/Map.hs

```diff
@@ -0,0 +1,175 @@
+    , symbolVerifiers
+    , builtinFunctions
+    ) where
+
+import qualified Data.HashMap.Strict as HashMap
+import           Data.Map
+
+import           Kore.AST.Common
+import           Kore.AST.MetaOrObject
+import           Kore.AST.PureML
+import           Kore.ASTUtils.SmartPatterns
+import qualified Kore.Builtin.Builtin as Builtin
+import           Kore.Builtin.Hook
+import           Kore.IndexedModule.MetadataTools
+import           Kore.Predicate.Predicate
+import           Kore.Step.ExpandedPattern
+import           Kore.Step.Function.Data
+import           Kore.Step.OrOfExpandedPattern
+import           Kore.Step.Simplification.Data
+import           Kore.Step.StepperAttributes
+
+
+{- | Builtin name of the @Map@ sort.
+ -}
+sort :: String
+sort = "MAP.Map"
+
+{- | Verify that the sort is hooked to the builtin @Int@ sort.
+
+  See also: 'sort', 'Builtin.verifySort'
+
+ -}
+assertSort :: Builtin.SortVerifier
+assertSort findSort = Builtin.verifySort findSort sort
+
+{- | Verify that hooked sort declarations are well-formed.
+
+  See also: 'Builtin.verifySortDecl'
+
+ -}
+sortDeclVerifiers :: Builtin.SortDeclVerifiers
+sortDeclVerifiers = HashMap.fromList [ (sort, Builtin.verifySortDecl) ]
+
+{- | Verify that hooked symbol declarations are well-formed.
+
+  See also: 'Builtin.verifySymbol'
+
+ -}
+symbolVerifiers :: Builtin.SymbolVerifiers
+symbolVerifiers =
+    HashMap.fromList
+    [
+      ("MAP.bind"
+      , Builtin.verifySymbol trivialVerifier
+          [ trivialVerifier
+          , trivialVerifier
+          , trivialVerifier
+          ]
+      )
+    , ("MAP.emp"
+      , Builtin.verifySymbol trivialVerifier
+          [ trivialVerifier
+          , trivialVerifier
+          , trivialVerifier
+          ]
+      )
+    , ("MAP.merge"
+      , Builtin.verifySymbol trivialVerifier
+          [ trivialVerifier
+          , trivialVerifier
+          ]
+      )
+    , ("MAP.element"
+      , Builtin.verifySymbol trivialVerifier
+          [ trivialVerifier
+          , trivialVerifier
+          ]
+      )
+    , ("MAP.lookup"
+      , Builtin.verifySymbol trivialVerifier
+          [ assertSort
+          , trivialVerifier
+          ]
+      )
+    ]
+  where
+    trivialVerifier :: Builtin.SortVerifier
+    trivialVerifier = const $ const $ Right ()
+
+isHook
+    :: MetadataTools Object StepperAttributes
+    -> SymbolOrAlias Object
+    -> String
+    -> Bool
+isHook tools sym hookName =
+    hook (attributes tools sym) == Hook (Just hookName)
+
+trivialExpandedPattern
+    :: MetaOrObject level
+    => PureMLPattern level var
+    -> ExpandedPattern level var
+trivialExpandedPattern p =
+    ExpandedPattern p makeTruePredicate []
+
+trivialEvalResult
+    :: (Applicative f, MetaOrObject level1)
+    => PureMLPattern level1 variable
+    -> f (AttemptedFunction level1 variable, SimplificationProof level2)
+trivialEvalResult p =
+    pure (Applied (MultiOr [trivialExpandedPattern p]), SimplificationProof)
+
+failedToEval
```

Comment:
We should also move `failedToEval` to `Kore.Builtin.Builtin` so it is only defined once.